### PR TITLE
[BLOCKING][jvm-packages] fix non-deterministic order within a partition  (in the case of an upstream shuffle) on prediction 

### LINKED ISSUE #4387: [jvm-packages] bug of XGBoostModel.transformInternal
ZipPartitions action at XGBoostModel.transformInternal will cause a mismatch between originalColumns and resultColumns.

`val resultRDD = inputRDD.zipPartitions(predictionRDD, preservesPartitioning = true) {
 `

In here, inputRDD and predictionRDD will be evaluated separated. Co-dependencies of these two RDDs will be computed twice, which may produce different row orders inside each partition if previous actions are not strictly keeping the same order (like ShuffleRead).

There is a tiny test below, which will fail because of column mismatching : 

> 

    val source = ss.sparkContext.parallelize((1 to 20).toList)
    val rnd = source.mapPartitions { iter =>
      iter.map(x => x -> Random.nextDouble())
        .toList
        .sortBy(_._2)
        .iterator
    }
    val negative = rnd.map { case (i, x) => i -> -x }
    val zipped = rnd.zip(negative)
    zipped.collect().foreach { case (x, y) =>
      x._1 shouldEqual y._1
    }

### PR DESCRIPTION
**Pull request by sperlingxx**
fix issue #4387 by replacing zip RDDs with caching original data in closure.

Corresponding unit tests have been added.

closes #4387 

closes https://github.com/dmlc/xgboost/issues/4307

### CODE REVIEW

**Comment by CodingCat** [comment]
thanks, yeah, this is a bug which is not easy to fix

this PR actually falls back to the previous problem about memory footprint, you can check https://github.com/dmlc/xgboost/commit/f368d0de2b6d60cdba67fed8291e7a3ce402850d#diff-a435450e9c28607f848ccf3246944a44 

let me think about what is the right way to fix, if we go to sort ,we need to do significant perf benchmarking before merging

**Comment by sperlingxx** [comment]
@CodingCat 
Thank you for replying me that quickly. 
Actually, this solution is quite like v0.81's implementation.
I have several ideas to improve this(maybe not good ideas = =)
1. do prediction in _MiniBatch_ style
    Maybe we can add a param like batch size?
2. add params _idColumn_ and _appendColumns_
   We filter other columns to reduce memory overhead of caching original data.

**Comment by CodingCat** [comment]
@sperlingxx can you elaborate more on the second approach?

**Comment by CodingCat** [comment]
at the same time I am benchmarking what if we sortWithinPartition beforehand

**Comment by CodingCat** [comment]
I tested with a prototype to sort each partition before hand, 

without sorting, we need 6+mins to finish the prediction over 120G input

with sort, it grows to 12+mins to finish the same task

**Comment by CodingCat** [comment]
I am experimenting with several potential solutions and finding more problems in our implementation, will update soon

**Comment by sperlingxx** [comment]
@CodingCat 
The second approach, we keep as fewer columns as we can, before the batch prediction stage(transformInternal).  So, we can cache fewer data of original DataFrame(inputRDD).

 something like:
>   dataset.toDF().select($(appendColumns) : _*)

**Comment by CodingCat** [comment]
ok, so I essentially tried three approaches to resolve the issue and finding more problems in XGBoost. 

0. approaches

I tried <b>sorting</b> the DataFrame before feeding to transformInternal(), <b>duplicate</b> dataset like the implementation here and <b>miniBatch</b>

1. benchmark of different approaches

I trained a model based on an internal dataset having 1.5~b rows and 20~ features and load the model to predict the training dataset in a separate Spark application.

To scale the test, I manually duplicate the dataset in Spark application and the benchmark results only counts the time spent on prediction stage

used code: https://github.com/CodingCat/xgboost4j-spark-scalability/blob/master/src/main/scala/me/codingcat/xgboost4j/PureXGBoostPredictor.scala

2. benchmark results

image

resources I used `--num-executors 100 --executor-memory 14g --executor-cores 8`

3. problems here, 

booster's prediction method is and has to be an synchronized method

**Comment by CodingCat** [comment]
the actions here

1. I will submit a PR to @sperlingxx 's branch soon 

2. fix the sharing issue in booster after this is merged (not blocking tho)

**Comment by hcho3** [comment]
@CodingCat BTW, you have permission to directly modify all PRs as a maintainer

**Comment by CodingCat** [comment]
@hcho3 how to modify with a bunch of changes?

**Comment by hcho3** [comment]
@CodingCat You can get a local clone of this PR by running
```
git clone --recursive https://github.com/sperlingxx/xgboost -b hot_fix_spark_estimator
```

**Comment by CodingCat** [comment]
I see.......thx

**Comment by sperlingxx** [comment]
@CodingCat 
 I'm not sure whether the implementation here will work in a minibatch way with the whole pipeline. 
So, I rewrote the classification part in a pure lazy(iterator) style. I hope it will be helpful :)

**Comment by CodingCat** [comment]
@sperlingxx can you explain how your implementation is a `pure lazy(iterator) style` and why my suggested implementation is not?

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassifier.scala [hunk 1]
```diff
@@ -287,46 +284,54 @@ class XGBoostClassificationModel private[ml](
...
-          if ($(useExternalMemory)) {
-            s"$appName-${TaskContext.get().stageId()}-dtest_cache-${TaskContext.getPartitionId()}"
-          } else {
-            null
+    val resultRDD = dataset.asInstanceOf[Dataset[Row]].rdd.mapPartitions { rowIterator =>
+      new AbstractIterator[Row] {
+        private var batchCnt = 0
+
+        private val batchIterImpl = rowIterator.grouped(
+          XGBoostClassificationModel.PREDICTION_BATCH_SIZE).flatMap { batchRow =>
```
Comment (CodingCat) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassifier.scala]: batchRow has been a Seq[Row] instead of Iterator[] here, so it's not lazy evaluated and stays in memory until this batch is finished

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassifier.scala [hunk 2]
```diff
@@ -287,46 +284,54 @@ class XGBoostClassificationModel private[ml](
...
+          if (batchCnt == 0) {
+            val rabitEnv = Array("DMLC_TASK_ID" -> TaskContext.getPartitionId().toString).toMap
+            Rabit.init(rabitEnv.asJava)
+          }
+
+          val features = batchRow.map(row => row.getAs[Vector]($(featuresCol)))
```
Comment (CodingCat) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassifier.scala]: regarding the memory footprint, you have put two Seq, one for Seq[Row], one for Seq[Vector] in memory, 

you can compare with my implementation, it only keeps a Seq[Row] due to iterator.duplicate()

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassifier.scala [hunk 3]
```diff
@@ -287,46 +284,54 @@ class XGBoostClassificationModel private[ml](
...
-          Rabit.shutdown()
-          Iterator(rawPredictionItr, probabilityItr, predLeafItr,
-            predContribItr)
-        } finally {
-          dm.delete()
+
+        override def hasNext: Boolean = batchIterImpl.hasNext
+
+        override def next(): Row = {
+          val ret = batchIterImpl.next()
```
Comment (CodingCat) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassifier.scala]: ah, I see, the key point is here, IIUC, with this implementation you don't need to locate prediction results in memory for a complete partition,

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassifier.scala (hunk 1 above)
Comment (CodingCat) [reply jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassifier.scala]: but we should think about the memory footprint in this place as the "grouped iterator" has been put in memory for twice

**Comment by CodingCat** [comment]
this is a significant change regarding performance and we need to be very careful about the correctness as well

can you also use internal dataset for evaluation?

**Comment by CodingCat** [comment]
ok, here is my latest benchmark results, looks like with the current implementation is slower than the diff I pushed yesterday

image


My theory is still on the synchronized method in Booster

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassifier.scala (hunk 3 above)
Comment (sperlingxx) [reply jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassifier.scala]: Yes. And, I've tried to mock your solution with the below script. 

```scala
object MiniBatchDebug {
  import scala.collection.Iterator
  import scala.collection.mutable.ListBuffer

  private val batchSize = 3
  private val dataSize = 9

  def f(iter: Iterator[Array[Int]]): Iterator[(Array[Int], Int)] = {
    var batchCnt = 0
    val results = new ListBuffer[Iterator[(Array[Int], Int)]]
    while (iter.hasNext) {
      println(s"fetch batch num: $batchCnt")
      val (x, y) = iter.take(batchSize).duplicate
      // mock the reading behavior of batch compute
      val ret = y.toArray.map(_.sum).iterator
      results += x.zip(ret)
      batchCnt += 1
    }
    results.reduce((i1, i2) => i1 ++ i2)
  }

  def main(args: Array[String]): Unit = {
    val dataSource = Seq.tabulate(dataSize)(x => Array(x, x * 2))
    val sourceIter: Iterator[Array[Int]] = dataSource.iterator
    val result = f(sourceIter).map { case (x, y) => x.sum + y }
    (1 to dataSize).foreach { i =>
      println(s"fetch element $i")
      result.next()
    }
  }
}
```

Here is the print info, which demonstrated it will run all batches before real data fetching :
>  fetch batch num: 0
fetch batch num: 1
fetch batch num: 2
fetch element 1
fetch element 2
fetch element 3
fetch element 4
fetch element 5
fetch element 6
fetch element 7
fetch element 8
fetch element 9

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassifier.scala (hunk 2 above)
Comment (sperlingxx) [reply jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassifier.scala]: Oh, it's an unnecessary footprint. Maybe it can be replaced by 
```scala
val features = batchRow.iterator.map(row => row.getAs[Vector]($(featuresCol)))
```

**Comment by sperlingxx** [comment]
> ok, here is my latest benchmark results, looks like with the current implementation is slower than the diff I pushed yesterday
> 
> image
> 
> My theory is still on the synchronized method in Booster

Thanks for benchmarking!

I'm a little confused about the __context switches__ caused by _synchronized prediction_ . Is it because there are multiple spark tasks running on each executor concurrently, and their booster share the same handle?

**Comment by CodingCat** [comment]
Yes, because we are using a broadcast booster which is singleton per executor, and regarding why we use broadcasted booster, you can check my previous comments

File: jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassifierSuite.scala [hunk 4]
```diff
@@ -27,11 +27,12 @@ import org.apache.spark.Partitioner
 
 class XGBoostClassifierSuite extends FunSuite with PerTest {
 
-  test("XGBoost-Spark XGBoostClassifier ouput should match XGBoost4j") {
+  test("XGBoost-Spark XGBoostClassifier output should match XGBoost4j") {
     val trainingDM = new DMatrix(Classification.train.iterator)
     val testDM = new DMatrix(Classification.test.iterator)
     val trainingDF = buildDataFrame(Classification.train)
     val testDF = buildDataFrame(Classification.test)
+    val randSortedTestDF = buildDataFrameWithRandSort(Classification.test)
```
Comment (CodingCat) [inline jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassifierSuite.scala]: let's separate them into two test to highlight the randsorting version and normal version of the test

File: jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostRegressorSuite.scala [hunk 5]
```diff
@@ -25,11 +25,12 @@ import org.scalatest.FunSuite
 
 class XGBoostRegressorSuite extends FunSuite with PerTest {
 
-  test("XGBoost-Spark XGBoostRegressor ouput should match XGBoost4j: regression") {
+  test("XGBoost-Spark XGBoostRegressor output should match XGBoost4j: regression") {
```
Comment (CodingCat) [inline jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostRegressorSuite.scala]: make the test name consistent with classifier part

File: jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/Booster.java [hunk 6]
```diff
@@ -47,7 +47,7 @@
    *                  the prediction of these DMatrices will become faster than not-cached data.
    * @throws XGBoostError native error
    */
-  Booster(Map<String, Object> params, DMatrix[] cacheMats) throws XGBoostError {
+  public Booster(Map<String, Object> params, DMatrix[] cacheMats) throws XGBoostError {
```
Comment (CodingCat) [inline jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/Booster.java]: this might no be necessary, maybe my bad

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassifier.scala [hunk 7]
```diff
@@ -287,46 +284,58 @@ class XGBoostClassificationModel private[ml](
...
+    val resultRDD = dataset.asInstanceOf[Dataset[Row]].rdd.mapPartitions { rowIterator =>
+      new AbstractIterator[Row] {
+        private var batchCnt = 0
+
+        private val batchIterImpl = rowIterator.grouped(
+          XGBoostClassificationModel.PREDICTION_BATCH_SIZE).flatMap { batchRow =>
+<<<<<<< HEAD
+=======
+
+>>>>>>> regressor impl
```
Comment (CodingCat) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassifier.scala]: my bad

**Comment by CodingCat** [comment]
I left more comments there, @sperlingxx would you please move forward with the PR

my suggestion is, use your way for the next version, and look at how to resolve shared properties among boosters after that

File: jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassifierSuite.scala (hunk 4 above)
Comment (sperlingxx) [reply jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassifierSuite.scala]: okay!

**Comment by CodingCat** [comment]
LGTM, thanks, will merge after CI is happy
